src/source/source_cache.js

```diff
@@ -241,14 +241,26 @@ class SourceCache extends Evented {
             tile.state = 'errored';
             if ((err: any).status !== 404) this._source.fire(new ErrorEvent(err, {tile}));
             // continue to try loading parent/children tiles if a tile doesn't exist (404)
-            else this.update(this.transform);
+            else {
+                const updateForTerrain = this._source.type === 'raster-dem' && this.usedForTerrain;
+                if (!updateForTerrain) {
+                    this.update(this.transform);
+                    return;
+                }
+
+                const painter = this.map ? this.map.painter : null;
+                const terrain = painter ? painter.terrain : null;
+                const tileSize = terrain ? terrain.getScaledDemTileSize() : this._source.tileSize;
+                this.update(this.transform, tileSize, updateForTerrain);
+                if (terrain) terrain.resetTileLookupCache();
```

Comment:
When `this.usedForTerrain` is true, should we assert that terrain exists? I don't think it's possible to hit this code path (when useForTerrain = true) without terrain.

```js
if (this.map && this.map.painter && this.map.painter.terrain) {
    this.update(this.transform, this.map.painter.terrain.getScaledDemTileSize(), true);
    this.map.painter.terrain.resetTileLookupCache();
}
```